File: src/main/java/io/github/Earth1283/fixLag/managers/PanicModeManager.java
package io.github.Earth1283.fixLag.managers;

import io.github.Earth1283.fixLag.FixLag;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Mob;
import org.bukkit.entity.Tameable;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

public class PanicModeManager extends BukkitRunnable {

    private final FixLag plugin;
    private final ConfigManager config;
    private final MessageManager messageManager;
    private boolean isPanicModeActive = false;
    private final Set<UUID> frozenEntities = new HashSet<>();

    public PanicModeManager(FixLag plugin) {
        this.plugin = plugin;
        this.config = plugin.getConfigManager();
        this.messageManager = plugin.getMessageManager();
    }

    @Override
    public void run() {
        if (!config.isPanicModeEnabled()) return;

        double tps = Bukkit.getServer().getTPS()[0];

        if (!isPanicModeActive && tps < config.getPanicModeTpsThreshold()) {
            activatePanicMode();
        } else if (isPanicModeActive && tps >= config.getPanicModeRecoverTps()) {
            deactivatePanicMode();
        }
    }

    private void activatePanicMode() {
        isPanicModeActive = true;
        int frozenCount = 0;

        for (World world : Bukkit.getWorlds()) {
            for (LivingEntity entity : world.getLivingEntities()) {
                if (entity instanceof Mob) {
                    Mob mob = (Mob) entity;
                    if (mob.getCustomName() != null) continue;
                    if (mob instanceof Tameable && ((Tameable) mob).isTamed()) continue;
                    
                    if (mob.isAware()) {
                        mob.setAware(false);
                        frozenEntities.add(mob.getUniqueId());
                        frozenCount++;
                    }
                }
            }
        }

        plugin.getLogger().warning("[Panic Mode] Activated! TPS dropped below threshold. Froze " + frozenCount + " mobs to save CPU.");
    }

    private void deactivatePanicMode() {
        isPanicModeActive = false;
        int thawedCount = 0;

        for (World world : Bukkit.getWorlds()) {
            for (LivingEntity entity : world.getLivingEntities()) {
                if (frozenEntities.contains(entity.getUniqueId())) {
                    if (entity instanceof Mob) {
                        ((Mob) entity).setAware(true);
                        thawedCount++;
                    }
                }
            }
        }
        
        frozenEntities.clear();
        plugin.getLogger().info("[Panic Mode] Deactivated! TPS recovered. Unfroze " + thawedCount + " mobs.");
    }
}

File: src/main/kotlin/io/github/Earth1283/fixLag/managers/PanicModeManager.kt
package io.github.Earth1283.fixLag.managers

import io.github.Earth1283.fixLag.FixLag
import org.bukkit.Bukkit
import org.bukkit.entity.Mob
import org.bukkit.entity.Tameable
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.world.ChunkUnloadEvent
import org.bukkit.scheduler.BukkitRunnable
import java.util.*

class PanicModeManager(private val plugin: FixLag) : BukkitRunnable(), Listener {

    private val config: ConfigManager = plugin.configManager
    private val messageManager: MessageManager = plugin.messageManager
    private var isPanicModeActive = false
    private val frozenEntities = mutableSetOf<UUID>()

    val isPanicActive: Boolean get() = isPanicModeActive
    val frozenMobCount: Int get() = frozenEntities.size

    override fun run() {
        if (!config.isPanicModeEnabled) return

        val tps = Bukkit.getServer().tps[0]

        if (!isPanicModeActive && tps < config.panicModeTpsThreshold) {
            activatePanicMode()
        } else if (isPanicModeActive && tps >= config.panicModeRecoverTps) {
            deactivatePanicMode()
        }
    }

    private fun activatePanicMode() {
        isPanicModeActive = true
        var frozenCount = 0

        for (world in Bukkit.getWorlds()) {
            for (entity in world.livingEntities) {
                if (entity is Mob) {
                    if (entity.getCustomName() != null) continue
                    if (entity is Tameable && entity.isTamed) continue

                    if (entity.isAware) {
                        entity.isAware = false
                        frozenEntities.add(entity.uniqueId)
                        frozenCount++
                    }
                }
            }
        }

        messageManager.logWarn("log_panic_activated", "<count>", frozenCount.toString())
    }

    private fun deactivatePanicMode() {
        isPanicModeActive = false
        var thawedCount = 0

        for (world in Bukkit.getWorlds()) {
            for (entity in world.livingEntities) {
                if (entity.uniqueId in frozenEntities) {
                    if (entity is Mob) {
                        entity.isAware = true
                        thawedCount++
                    }
                }
            }
        }

        frozenEntities.clear()
        messageManager.logInfo("log_panic_deactivated", "<count>", thawedCount.toString())
    }

    // Restore AI on frozen mobs before the chunk is saved to disk so NoAI doesn't persist.
    @EventHandler(ignoreCancelled = true)
    fun onChunkUnload(event: ChunkUnloadEvent) {
        if (!isPanicModeActive || frozenEntities.isEmpty()) return
        for (entity in event.chunk.entities) {
            if (entity.uniqueId in frozenEntities) {
                if (entity is Mob) entity.isAware = true
                frozenEntities.remove(entity.uniqueId)
            }
        }
    }
}
